mypy/build.py

```diff
@@ -778,95 +785,188 @@ def is_file(path: str) -> bool:
+        self.namespaces_allowed = namespaces_allowed
+        self._find_module_cache = {}  # type: Dict[str, List[str]]
+
+    def find_module(self, id: str) -> Optional[str]:
+        paths = self._find_module(id)
+        if paths:
+            return paths[0]
+        else:
+            return None
+
+    def find_modules_recursive(self, module: str) -> List[BuildSource]:
+        """
+        Discover module and all it's children
+        Remove duplicates from discovered paths
+        """
+        hits = set()  # type: Set[str]
+        result = []  # type: List[BuildSource]
+        for src in self._find_modules_recursive(module):
+            if src.module not in hits:
+                hits.add(src.module)
+                result.append(src)
+        return result
+
+    def _find_modules_recursive(self, module: str) -> List[BuildSource]:
+        module_paths = self._find_module(module)
+
+        srcs = []  # type: List[BuildSource]
+        for path in module_paths:
+            if is_module_path(path) or is_pkg_path(path):
+                srcs.append(BuildSource(path, module, None))
+
+            if is_pkg_path(path):
+                path = dirname(path)
+                for submodule in self._find_submodules(module, path):
+                    srcs += self._find_modules_recursive(submodule)
+            elif is_namespace_path(path):
+                for submodule in self._find_submodules(module, path):
+                    srcs += self._find_modules_recursive(submodule)
+
+        return srcs
+
+    def _find_submodules(self, module: str, path: str) -> Iterator[str]:
+        for item in list_dir(path) or []:
+            if item == '__init__.py' or item == '__init__.pyi':
+                continue
+
+            if item.endswith(tuple(PYTHON_EXTENSIONS)):
+                item = item.split('.')[0]
+
+            yield module + '.' + item
+
+    def _collect_paths(self, paths: List[str], last_comp: str) -> List[str]:
+        """
+        Collect all available module paths
+        """
+        sepinit = '__init__'
+        ctx = ImportContext()
+
+        # Detect modules in following order: package, module, namespace.
+        # First hit determines module type, consistency of paths to given type
+        # ensured in ImportContext
+        for path_item in paths:
+            if is_module_path(path_item):
+                continue
+
+            if is_pkg_path(path_item):
+                path_item = dirname(path_item)
+
+            for ext in PYTHON_EXTENSIONS:
+                path = os.path.join(path_item, last_comp, sepinit + ext)
+                ctx.maybe_add_path(path, ModuleType.package)
+
+            for ext in PYTHON_EXTENSIONS:
+                path = os.path.join(path_item, last_comp + ext)
+                ctx.maybe_add_path(path, ModuleType.module)
+
+        if self.namespaces_allowed and not ctx.paths:
+            for path_item in paths:
```

Comment:
rather than repeating the loop and the `is_pkg_path` check, I'd recommend building up the list of "possible namespace package paths" in its own list variable on your first iteration through `paths`, and then just using that already-collected list here.